File: backend/app/api/ws.py
"""WebSocket для интерактивного чата в ветках: мгновенная доставка + «печатает…».

Работает с одним воркером uvicorn (менеджер соединений хранится в процессе).
"""
from collections import defaultdict

import jwt
from fastapi import APIRouter, Query, WebSocket, WebSocketDisconnect
from sqlalchemy import func

from app.api.routes.threads import _accessible, _mark_read, _mark_staff_seen
from app.core.database import SessionLocal
from app.core.security import decode_access_token
from app.models import Thread, ThreadMessage, User

router = APIRouter()


class Manager:
    def __init__(self):
        self.rooms: dict[int, set[WebSocket]] = defaultdict(set)

    async def broadcast(self, thread_id: int, data: dict):
        for ws in list(self.rooms[thread_id]):
            try:
                await ws.send_json(data)
            except Exception:
                self.rooms[thread_id].discard(ws)


manager = Manager()


def _auth(token: str) -> User | None:
    try:
        email = decode_access_token(token).get("sub")
    except jwt.PyJWTError:
        return None
    if not email:
        return None
    with SessionLocal() as db:
        return db.query(User).filter(User.email == email).first()


@router.websocket("/ws/threads/{thread_id}")
async def ws_thread(websocket: WebSocket, thread_id: int, token: str = Query(...)):
    user = _auth(token)
    if not user or not user.is_active:
        await websocket.close(code=1008)
        return
    with SessionLocal() as db:
        if not _accessible(db, user).filter(Thread.id == thread_id).first():
            await websocket.close(code=1008)
            return

    await websocket.accept()
    manager.rooms[thread_id].add(websocket)
    try:
        while True:
            data = await websocket.receive_json()
            typ = data.get("type")
            if typ == "typing":
                await manager.broadcast(thread_id, {"type": "typing", "user_id": user.id, "name": user.full_name})
            elif typ == "message":
                body = (data.get("body") or "").strip()
                if not body:
                    continue
                with SessionLocal() as db:
                    if not _accessible(db, user).filter(Thread.id == thread_id).first():
                        continue
                    reply_to_id = None
                    rid = data.get("reply_to_id")
                    if rid:
                        parent = db.get(ThreadMessage, rid)
                        if parent and parent.thread_id == thread_id:
                            reply_to_id = parent.id
                    msg = ThreadMessage(thread_id=thread_id, author_id=user.id, body=body, reply_to_id=reply_to_id)
                    db.add(msg)
                    t = db.get(Thread, thread_id)
                    t.updated_at = func.now()
                    _mark_staff_seen(db, user, t)
                    db.commit()
                    db.refresh(msg)
                    _mark_read(db, user, thread_id)
                    from app.api.routes.threads import _reply_dict
                    payload = {
                        "type": "message",
                        "message": {
                            "id": msg.id, "author_id": user.id, "author_name": user.full_name,
                            "body": msg.body, "created_at": msg.created_at.isoformat(),
                            "edit_count": 0, "reactions": [], "reply_to": _reply_dict(msg),
                        },
                    }
                await manager.broadcast(thread_id, payload)
    except WebSocketDisconnect:
        manager.rooms[thread_id].discard(websocket)
    except Exception:
        manager.rooms[thread_id].discard(websocket)

File: backend/internal/web/ws.go
package web

import (
	"net/http"
	"strconv"
	"strings"
	"sync"

	"github.com/go-chi/chi/v5"
	"github.com/gorilla/websocket"
	"gorm.io/gorm"

	"manibandha/internal/models"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true }, // токен в query аутентифицирует
}

// wsConn — обёртка с мьютексом (конкурентная запись в gorilla-сокет небезопасна).
type wsConn struct {
	ws *websocket.Conn
	mu sync.Mutex
}

func (c *wsConn) writeJSON(v any) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.ws.WriteJSON(v)
}

// ── Хаб чата: один пользователь → набор сокетов ──────────────────────────────

type chatHubT struct {
	mu      sync.RWMutex
	sockets map[int]map[*wsConn]bool
}

var chatH = &chatHubT{sockets: map[int]map[*wsConn]bool{}}

func (h *chatHubT) add(uid int, c *wsConn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.sockets[uid] == nil {
		h.sockets[uid] = map[*wsConn]bool{}
	}
	h.sockets[uid][c] = true
}

func (h *chatHubT) remove(uid int, c *wsConn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if m := h.sockets[uid]; m != nil {
		delete(m, c)
		if len(m) == 0 {
			delete(h.sockets, uid)
		}
	}
}

func (h *chatHubT) sendToUsers(uids []int, data any) {
	h.mu.RLock()
	seen := map[*wsConn]bool{}
	var targets []*wsConn
	for _, uid := range uids {
		for c := range h.sockets[uid] {
			if !seen[c] {
				seen[c] = true
				targets = append(targets, c)
			}
		}
	}
	h.mu.RUnlock()
	for _, c := range targets {
		_ = c.writeJSON(data)
	}
}

func (s *Server) broadcastChat(chatID int, data any) {
	ids := s.memberIDs(chatID)
	go chatH.sendToUsers(ids, data)
}

// онлайн, если у пользователя есть хотя бы один активный чат-сокет
func (h *chatHubT) isOnline(uid int) bool {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return len(h.sockets[uid]) > 0
}

// ── Хаб веток: тема → набор сокетов ──────────────────────────────────────────

type threadHubT struct {
	mu    sync.RWMutex
	rooms map[int]map[*wsConn]bool
}

var threadH = &threadHubT{rooms: map[int]map[*wsConn]bool{}}

func (h *threadHubT) add(tid int, c *wsConn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.rooms[tid] == nil {
		h.rooms[tid] = map[*wsConn]bool{}
	}
	h.rooms[tid][c] = true
}

func (h *threadHubT) remove(tid int, c *wsConn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if m := h.rooms[tid]; m != nil {
		delete(m, c)
		if len(m) == 0 {
			delete(h.rooms, tid)
		}
	}
}

func (h *threadHubT) broadcast(tid int, data any) {
	h.mu.RLock()
	var targets []*wsConn
	for c := range h.rooms[tid] {
		targets = append(targets, c)
	}
	h.mu.RUnlock()
	for _, c := range targets {
		_ = c.writeJSON(data)
	}
}

func (s *Server) broadcastThread(threadID int, data any) {
	go threadH.broadcast(threadID, data)
}

// ── аутентификация WS по токену ──────────────────────────────────────────────

func (s *Server) wsAuth(token string) *models.User {
	claims, err := s.JWT.Parse(token)
	if err != nil {
		return nil
	}
	email, _ := claims["sub"].(string)
	if email == "" {
		return nil
	}
	var u models.User
	if err := s.DB.Where("email = ?", email).First(&u).Error; err != nil || !u.IsActive {
		return nil
	}
	return &u
}

// GET /ws/chat?token=...
func (s *Server) wsChat(w http.ResponseWriter, r *http.Request) {
	u := s.wsAuth(r.URL.Query().Get("token"))
	if u == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	ws, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		return
	}
	c := &wsConn{ws: ws}
	chatH.add(u.ID, c)
	defer func() {
		chatH.remove(u.ID, c)
		ws.Close()
	}()
	for {
		var data map[string]any
		if err := ws.ReadJSON(&data); err != nil {
			return
		}
		if data["type"] == "typing" {
			cid, ok := toInt(data["chat_id"])
			if !ok {
				continue
			}
			var cnt int64
			s.DB.Model(&models.ChatMember{}).Where("chat_id = ? AND user_id = ?", cid, u.ID).Count(&cnt)
			if cnt == 0 {
				continue
			}
			others := []int{}
			for _, id := range s.memberIDs(cid) {
				if id != u.ID {
					others = append(others, id)
				}
			}
			chatH.sendToUsers(others, map[string]any{
				"type": "typing", "chat_id": cid, "user_id": u.ID, "name": u.FullName,
			})
		}
	}
}

// GET /ws/threads/{id}?token=...
func (s *Server) wsThread(w http.ResponseWriter, r *http.Request) {
	u := s.wsAuth(r.URL.Query().Get("token"))
	if u == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	tid, _ := strconv.Atoi(chi.URLParam(r, "id"))
	if _, ok := s.accessibleThread(u, tid); !ok {
		http.Error(w, "forbidden", http.StatusForbidden)
		return
	}
	ws, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		return
	}
	c := &wsConn{ws: ws}
	threadH.add(tid, c)
	defer func() {
		threadH.remove(tid, c)
		ws.Close()
	}()
	for {
		var data map[string]any
		if err := ws.ReadJSON(&data); err != nil {
			return
		}
		switch data["type"] {
		case "typing":
			threadH.broadcast(tid, map[string]any{"type": "typing", "user_id": u.ID, "name": u.FullName})
		case "message":
			body, _ := data["body"].(string)
			body = strings.TrimSpace(body)
			if body == "" {
				continue
			}
			if _, ok := s.accessibleThread(u, tid); !ok {
				continue
			}
			var replyTo *int
			if rid, ok := toInt(data["reply_to_id"]); ok && rid > 0 {
				var parent models.ThreadMessage
				if err := s.DB.First(&parent, rid).Error; err == nil && parent.ThreadID == tid {
					replyTo = &parent.ID
				}
			}
			msg := models.ThreadMessage{ThreadID: tid, AuthorID: &u.ID, Body: body, ReplyToID: replyTo}
			s.DB.Create(&msg)
			t := &models.Thread{ID: tid, Kind: threadKind(s, tid)}
			s.DB.Model(&models.Thread{}).Where("id = ?", tid).Update("updated_at", gorm.Expr("now()"))
			s.markStaffSeen(u, t)
			s.markRead(u.ID, tid)
			var full models.ThreadMessage
			s.DB.Preload("ReplyTo").Preload("ReplyTo.Author").First(&full, msg.ID)
			threadH.broadcast(tid, map[string]any{
				"type": "message",
				"message": map[string]any{
					"id": msg.ID, "author_id": u.ID, "author_name": u.FullName,
					"body": msg.Body, "created_at": tsUTC(msg.CreatedAt),
					"edit_count": 0, "reactions": []any{}, "reply_to": replyDict(&full),
				},
			})
		}
	}
}

func threadKind(s *Server, tid int) string {
	var t models.Thread
	s.DB.Select("kind").First(&t, tid)
	return t.Kind
}
